### PR TITLE
Add SQLite mixed-type data handling capabilities

### PR DESCRIPTION
**Pull request by d33bs**
# Description

This change seeks to address #38 through dynamic handling of mixed-type SQLite column data when exceptions of this kind are encountered with DuckDB SQLite queries.

When mixed type exceptions are encountered, the design intent here strives to keep as much data as possible while assuming one or many values of mismatching type within the SQLite table being queried. SQLite3 SQL case statements are formed to convert mismatched storage class values to NULL's for every column in the table being queried. These SQLite NULL's are interpreted by `sqlite3` as Python `None`'s, which are then interpreted as `pyarrow.null()`'s for export to Parquet.

For SQLite data sources where mixed typing exceptions are not encountered, the default capabilities through DuckDB function as usual.

## What is the nature of your change?

- [x] Bug fix (fixes an issue).
- [ ] Enhancement (adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).
- [ ] This change requires a documentation update.

# Checklist

Please ensure that all boxes are checked before indicating that a pull request is ready for review.

- [x] I have read the [CONTRIBUTING.md](CONTRIBUTING.md) guidelines.
- [x] My code follows the style guidelines of this project.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] New and existing unit tests pass locally with my changes.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have deleted all non-relevant text in this pull request template.

### CODE REVIEW

File: cytotable/utils.py
Comment (falquaddoomi) [inline cytotable/utils.py:174]: This is a very minor nitpick, but some argument descriptions end in a period and some don't. I don't have a preference between the two, but I think it'd be better if they were consistent.

File: cytotable/utils.py
Comment (falquaddoomi) [inline cytotable/utils.py:204]: Nifty! 🎉

File: tests/test_convert.py
Comment (falquaddoomi) [inline tests/test_convert.py:862]: Same comment here as before regarding using `duckdb.TypeMismatchException` over its ancestor `duckdb.Error`.

File: tests/conftest.py
Comment (falquaddoomi) [inline tests/conftest.py:546]: Cool that pytest uses generators to allow for cleanup steps after creating a fixture. I don't know if this is necessary since I'm not super familiar with pytest, but this might be more robust as a try/finally, e.g.:

```suggestion
    try:
        yield filepath
    finally:
        # after completing the tests, remove the file
        pathlib.Path(filepath).unlink()
```

That way, even if the coprocess to which you're yielding never returns (not that that's likely, but who knows?), you're guaranteed to delete the file. Again, this might be too cautious, so I'll leave it up to you to choose. (On a side note, I have no idea why github's syntax colorer isn't coloring "finally", but that's definitely valid Python code...)

File: cytotable/convert.py
Comment (d33bs) [reply cytotable/convert.py:171]: Thanks for calling this forward! I noticed mixed behavior with DuckDB Python Exceptions which resulted in the approach here. I'll give the more specific exception another attempt here (very much prefer it! 🙂 ).

File: cytotable/utils.py
Comment (d33bs) [reply cytotable/utils.py:174]: Good point, thanks, I'll update these!

File: cytotable/convert.py
Comment (d33bs) [reply cytotable/convert.py:171]: Digging into this more I found that only generic errors are raised from the context of when this happens from the `sqlite_scanner` extension. I'm going to raise an issue for that extension with the hope that we may get more feedback here. In the meantime, I'll plan to leave things as-is.

See the following demonstration code:

```python
import sqlite3

import duckdb

# create a temporary sqlite connection
filepath = ".example.sqlite"

# statements for creating database with simple structure
create_stmts = """
    DROP TABLE IF EXISTS tbl_a;
    CREATE TABLE tbl_a (
        col_integer INTEGER NOT NULL
        ,col_text TEXT
        ,col_blob BLOB
        ,col_real REAL
    );
    """

# some example values to insert into the database
insert_vals = ["nan", "sample", b"sample_blob", 0.5]

# create the database and insert some data into it
with sqlite3.connect(filepath) as connection:
    connection.executescript(create_stmts)

    connection.execute(
        (
            "INSERT INTO tbl_a (col_integer, col_text, col_blob, col_real)"
            "VALUES (?, ?, ?, ?);"
        ),
        insert_vals,
    )

try:
    duckdb.connect().execute(
        f"""
    /* install and load sqlite plugin for duckdb */
    INSTALL sqlite_scanner;
    LOAD sqlite_scanner;

    /* perform query on tbl_a table */
    SELECT * FROM sqlite_scan('{filepath}', 'tbl_a');
    """
    ).arrow()
    
except duckdb.TypeMismatchException as type_mismatch_e:
    print("Found a type mismatch exception: ", type_mismatch_e)
    
except Exception as e:
    print("Found a different error of type: ", type(e), e, sep="\n")
```

File: tests/test_convert.py
Comment (d33bs) [reply tests/test_convert.py:862]: Similar to the previous comment, leaving this as-is for right now.

File: cytotable/convert.py
Comment (falquaddoomi) [reply cytotable/convert.py:171]: Nice detective work, and a pity the extension isn't throwing the correct exception. Sounds good regarding leaving things as-is for now, since there aren't any better options.

**Comment by gwaybio** [comment]
@falquaddoomi @d33bs - is this ready to merge? Will it solve #38 (@jenna-tomkinson and @MattsonCam are currently blocked by it for the JUMP single cell analysis)

**Comment by gwaybio** [comment]
Update, @d33bs - just saw your comment in https://github.com/cytomining/CytoTable/issues/38#issuecomment-1625615593

**Review by gwaybio — APPROVED** [review]
LGTM! Excited to get this merged in.

**Pull request merged**
